packages/prisma-client/prisma/migrations/20240524155926_patch_map/migration.sql

```diff
@@ -0,0 +1,200 @@
+ * SOFTWARE.
+ */
+
+/**
+ * @typedef {Object} Patch
+ * @property {"replace"|"remove"|"add"} op - The operation to perform.
+ * @property {(string|number)[]} path - The path where the operation should be applied.
+ * @property {*} [value] - The value to be used with the operation (optional).
+ */
+
+/**
+ *
+ * @param {unknown} thing
+ * @returns
+ */
+const getArchtype = (thing) =>
+  Array.isArray(thing)
+    ? "Array"
+    : thing instanceof Map
+      ? "Map"
+      : thing instanceof Set
+        ? "Set"
+        : "Object";
+
+// @ts-ignore
+const get = (thing, prop) =>
+  getArchtype(thing) === "Map" ? thing.get(prop) : thing[prop];
+
+/**
+ * https://github.com/immerjs/immer/blob/e2d222bd4fb26abded04075c936290715e9ee335/src/plugins/patches.ts#L214
+ * @template T
+ * @param {T} draft - The draft object to apply patches to.
+ * @param {ReadonlyArray<Patch>} patches
+ * @returns {T} The patched draft object.
+ */
+const applyPatches = (draft, patches) => {
+  patches.forEach((patch) => {
+    const { path, op } = patch;
+
+    let base = draft;
+    for (let i = 0; i < path.length - 1; i++) {
+      // @ts-ignore
+      const parentType = getArchtype(base);
+      let p = path[i];
+      if (typeof p !== "string" && typeof p !== "number") {
+        p = "" + p;
+      }
+
+      base = get(base, p);
+
+      if (typeof base !== "object") {
+        throw new Error("die");
+      }
+    }
+
+    const type = getArchtype(base);
+    const value = patch.value;
+
+    const key = path[path.length - 1];
+    switch (op) {
+      case "replace":
+        switch (type) {
+          case "Map":
+            // @ts-ignore
+            return base.set(key, value);
+          case "Set":
+            throw new Error("set is not supported");
+          default:
+            // @ts-ignore
+            return (base[key] = value);
+        }
+      case "add":
+        switch (type) {
+          case "Array":
+            // @ts-ignore
+            return key === "-" ? base.push(value) : base.splice(key, 0, value);
+          case "Map":
+            // @ts-ignore
+            return base.set(key, value);
+          case "Set":
+            // @ts-ignore
+            return base.add(value);
+          default:
+            // @ts-ignore
+            return (base[key] = value);
+        }
+      case "remove":
+        switch (type) {
+          case "Array":
+            // @ts-ignore
+            return base.splice(key, 1);
+          case "Map":
+            // @ts-ignore
+            return base.delete(key);
+          case "Set":
+            // @ts-ignore
+            return base.delete(patch.value);
+          default:
+            // @ts-ignore
+            return delete base[key];
+        }
+      default:
+        throw new Error("invalid op");
+    }
+  });
+
+  return draft;
+};
+
+/**
+ *
+ * @param {string} dataStr
+ * @param {string} primaryKeyCommaSeparated - if empty then it's js array or plain object, otherwise is a map
```

Comment:
if we can avoid havinng multiple types here we should, btw also @param types are wrong and we are not using them anywhere else

Reply:
What means wrong, this is fully typecheked js file. We used JSDoc typescript types before in middleware.js
There are no other ways to have types in js

Reply:
And js file as described in comments copy pasted here